### PR TITLE
fix deadlock - don't accept keyPress when processing dbus call

### PR DESCRIPTION
**Pull request by technic**
should fix https://github.com/lxde/lxqt/issues/1301

### CODE REVIEW

File: daemon/core.cpp [hunk 1]
```diff
@@ -1146,17 +1156,18 @@ void Core::run()
...
                 break;
             }
 
-            switch (event.type)
-            {
-            case KeyPress:
+            if (event.type == KeyPress && mDataMutex.tryLock(0))
             {
-                QMutexLocker lock(&mDataMutex);
+                MutexUnlocker unlocker(&mDataMutex);
```
Comment (palinek) [inline daemon/core.cpp]: How about using the `unique_lock`? Consider this diff:
```diff
diff --git a/daemon/core.cpp b/daemon/core.cpp
index b267765..68e6efb 100644
--- a/daemon/core.cpp
+++ b/daemon/core.cpp
@@ -42,6 +42,7 @@
 #include <errno.h>
 #include <string.h>
 #include <stdio.h>
+#include <mutex>
 
 #include <stdexcept>
 
@@ -57,16 +58,6 @@
 #include "core.h"
 
 
-class MutexUnlocker
-{
-public:
-    MutexUnlocker(QMutex *mutex): m(mutex) {}
-    ~MutexUnlocker() { m->unlock(); }
-private:
-    QMutex *m;
-};
-
-
 enum
 {
     X11_OP_StringToKeycode,
@@ -1164,7 +1155,7 @@ void Core::run()
 
             if (event.type == KeyPress && mDataMutex.tryLock(0))
             {
-                MutexUnlocker unlocker(&mDataMutex);
+                std::unique_lock<QMutex> unlocker{mDataMutex, std::adopt_lock};
 
                 // pop event from the x11 queue and process it
                 XNextEvent(mDisplay, &event);
```

**Comment by stefonarch** [comment]
Does not change here - panel unresponsive and shortcuts not available as before.

**Comment by palinek** [comment]
> Does not change here - panel unresponsive and shortcuts not available as before.

Are you sure? With my humble test (stop lxqt-globalkeysd, run freshly built lxqt-globalkeysd from terminal and holding down some shortcut (e.g. alt+f1)) I'm not able to reproduce the freeze any more...

**Comment by technic** [comment]
@stefonarch 
Could you please show us your backtrace

**Comment by stefonarch** [comment]
I did 2 tests afteer compiling the branch, one restarting session, one restarting PC and hitting alt+p immediately (should open Pcmanfm here): in both cases no shortkeys, killing globalkeysd conduced to a completely blocked session but executed alt+p. 
For creating a backtrace I need help, I'm not that experienced

**Comment by technic** [comment]
> For creating a backtrace I need help, I'm not that experienced

In the frozen state use
```
gdb -p $(pidof lxqt-globalkeysd) -ex 'thread apply all bt'
```
Also you need to rebuild lxqt-globalkeysd with debug symbols, on arch there is a manual https://wiki.archlinux.org/index.php/Debug_-_Getting_Traces

**Comment by technic** [comment]
Is it only pcmanfm shortcut, that causes problems?

**Comment by stefonarch** [comment]
No, tested also with alt+tab, I compiled with -DCMAKE_BUILD_TYPE=Debug but your command produce this:
`
$ gdb -p $(pidof lxqt-globalkeysd) -ex 'thread apply all bt'
GNU gdb (GDB) 8.0.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
9404: File o directory non esistente.
Attaching to process 11132
ptrace: Operazione non permessa.
(gdb) while
if/while commands require arguments.
(gdb) `

**Comment by technic** [comment]
you need to allow attach to another process in kernel with
```
echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope
```

**Comment by stefonarch** [comment]
ok, here it is:

```
9404:` File o directory non esistente.
Attaching to process 12499
[New LWP 12518]
[New LWP 12539]
[New LWP 12548]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
0x00007fc3485717d8 in read () from /usr/lib/libc.so.6
                                                                                                                         
Thread 4 (Thread 0x7fc33127e700 (LWP 12548)):                                                                            
#0  0x00007fc34857abb9 in syscall () at /usr/lib/libc.so.6                                                               
#1  0x00007fc3491ce7a6 in QBasicMutex::lockInternal() () at /usr/lib/libQt5Core.so.5                                     
#2  0x0000559437b5d3c1 in QMutexLocker::QMutexLocker(QBasicMutex*) (this=0x7fc33127dc48, m=0x7ffeacc270a0)               
    at /usr/include/qt/QtCore/qmutex.h:200                                                                               
#3  0x0000559437b4f178 in Core::run() (this=0x7ffeacc27030) at /home/stef/git/lxqt-globalkeys/daemon/core.cpp:1428       
#4  0x00007fc3491d715b in  () at /usr/lib/libQt5Core.so.5                                                                
#5  0x00007fc34783a08a in start_thread () at /usr/lib/libpthread.so.0                                                    
#6  0x00007fc3485801bf in clone () at /usr/lib/libc.so.6                                                                 
                                                                                                                         
Thread 3 (Thread 0x7fc334170700 (LWP 12539)):
#0  0x00007fc348575cbb in poll () at /usr/lib/libc.so.6
#1  0x00007fc345a78c09 in  () at /usr/lib/libglib-2.0.so.0
#2  0x00007fc345a78d1c in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#3  0x00007fc349415061 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) ()
    at /usr/lib/libQt5Core.so.5
#4  0x00007fc3493b8ffb in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#5  0x00007fc3491d240e in QThread::exec() () at /usr/lib/libQt5Core.so.5
#6  0x00007fc34aeb1396 in  () at /usr/lib/libQt5DBus.so.5
#7  0x00007fc3491d715b in  () at /usr/lib/libQt5Core.so.5
```
Hitting immediately a shortcut the panel is still responsive for about 30-40 seconds, then a notification pops up saying that the shortcut for mixer cannot be registered, after that panel is blocked and I cannot even kill it with killall.

**Comment by technic** [comment]
> ok, here it is:
cool thanks! but you lost output for threads 1 and 2 :)

**Comment by technic** [comment]
Wait, @stefonarch I am suspecting you are on the _wrong branch_, you must use **fix-deadlock** from my fork https://github.com/technic/lxqt-globalkeys

**Comment by stefonarch** [comment]
O mess, in fact I compiled deadlock_workaround_test from master... your fix-deadlock works like a charm :)

**Comment by palinek** [comment]
> your fix-deadlock works like a charm :)

Good.

@technic after squashing... GTM

**Comment by agaida** [comment]
woulnd't squash and merge be enough?

**Comment by fulalas** [comment]
Nice one, @palinek! It's working nice :+1:

**Comment by palinek** [comment]
It wasn't me... @technic did the job

**Comment by fulalas** [comment]
Oops! My mistake. Sorry.

**Pull request merged**